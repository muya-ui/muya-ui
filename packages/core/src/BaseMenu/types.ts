import React from 'react';

import { IComponentSizeSpec } from '@muya-ui/theme-light';

import { ISizeSpecBaseProps } from '../types';

export interface IBaseMenuItemBaseProps extends ISizeSpecBaseProps {
  /**
   * item 的内容
   */
  children: React.ReactNode;
  /**
   * item 是否禁用
   */
  disabled?: boolean;
  /**
   * item 是否选中
   */
  selected?: boolean;
  /**
   * item 是否激活，配合键盘事件
   */
  active?: boolean;
  /**
   * item 是否加载中
   */
  loading?: boolean;
  /**
   * 加载的 icon
   */
  loadingIcon?: React.ReactNode;
  /**
   * 选中的 icon
   */
  selectedIcon?: React.ReactNode;
}

export type IBaseMenuItemProps = IBaseMenuItemBaseProps & React.HTMLAttributes<HTMLDivElement>;

export interface IBaseMenuItemGroupBaseProps {
  /**
   * item group 的内容
   */
  label: React.ReactNode;
  /**
   * item group 下的 item
   */
  children: React.ReactNode;
}

export type IBaseMenuItemGroupProps = IBaseMenuItemGroupBaseProps &
  React.HTMLAttributes<HTMLDivElement>;

export interface IBaseMenuBaseProps extends ISizeSpecBaseProps {
  /**
   * 菜单宽度
   */
  width?: number | Record<IComponentSizeSpec, number>;
  /**
   * item 超出多少滚动
   */
  maxItemCountPerPage?: number;
  /**
   * item 的内容
   */
  children: React.ReactElement[] | React.ReactElement;
  /**
   * 菜单滚动
   */
  onScroll?: React.UIEventHandler<HTMLDivElement>;
}

export type IBaseMenuProps = IBaseMenuBaseProps & React.HTMLAttributes<HTMLDivElement>;
