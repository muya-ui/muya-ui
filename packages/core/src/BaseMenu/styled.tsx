import styled, { css } from 'styled-components';

import { IComponentSizeSpec } from '@muya-ui/theme-light';

import ellipsisStyle from '../styled/mixins/ellipsis';
import { itemDisabledStyle, itemNormalStyle, itemSelectedStyle } from '../styled/mixins/item';
import { scrollBarStyle } from '../ScrollView/styled';
import { IThemedBaseProps } from '../types';

export interface IStyledBaseMenuProps extends IThemedBaseProps {
  $size?: IComponentSizeSpec;
  $width?: number;
  $maxItemCountPerPage?: number;
  $hasScrollBar: boolean;
}

export interface IStyledMenuItemProps extends IThemedBaseProps {
  $size?: IComponentSizeSpec;
  $active: boolean;
  $disabled: boolean;
  $selected: boolean;
}

export const menuItemCss = (props: IStyledMenuItemProps) => {
  const {
    $size,
    $active,
    $selected,
    $disabled,
    theme: {
      transition: {
        pattern: { duration, easing },
      },
    },
  } = props;
  const token = props.theme.components.BaseMenu.item;
  return css`
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex: 1;
    box-sizing: border-box;
    padding: 0 ${token.paddingHorizontal}px;
    font-size: ${token.fontSize[$size!]}px;
    line-height: ${token.height[$size!]}px;
    height: ${token.height[$size!]}px;
    cursor: pointer;
    transition: background-color ${duration.status}ms ${easing.status},
      color ${duration.status}ms ${easing.status};
    ${itemNormalStyle(token)}
    span {
      ${ellipsisStyle}
    }

    ${$active &&
      css`
        color: ${token.color.hover};
        background-color: ${token.background.hover};
        &:hover {
          color: ${token.color.hover};
          background-color: ${token.background.hover};
        }
        &:active {
          color: ${token.color.hover};
          background-color: ${token.background.hover};
        }
      `}
    ${$selected && itemSelectedStyle(token)}
    ${$disabled && itemDisabledStyle(token)}
  `;
};

export const StyledBaseMenuItem = styled.div`
  ${menuItemCss};
`;

export const StyledBaseMenuItemGroupLabel = styled.span`
  display: flex;
  justify-content: space-between;
  align-items: center;
  box-sizing: border-box;
  cursor: default;
`;

export const StyledBaseMenuItemGroup = styled.div``;

export const StyledBaseMenuItemDivider = styled.div`
  height: 1px;
`;

export const StyledBaseMenuWrapper = styled.div`
  overflow-y: auto;
`;
export const baseMenuCss = (props: IStyledBaseMenuProps) => {
  const {
    $size,
    $width,
    $hasScrollBar,
    $maxItemCountPerPage,
    theme: { colors, shadows },
  } = props;
  const token = props.theme.components.BaseMenu;
  const {
    wrapper: wrapperToken,
    item: itemToken,
    divider: dividerToken,
    group: groupToken,
  } = token;
  return css`
    box-shadow: ${shadows.pattern.popper.normal};
    background: ${colors.pattern.background.higher};
    border-radius: ${wrapperToken.borderRadius};
    padding: ${wrapperToken.paddingVertical}px 0;
    overflow: hidden;
    ${$width && `width: ${$width}px`};
    ${$hasScrollBar && `padding-right: ${wrapperToken.paddingRight}px`};

    ${StyledBaseMenuWrapper} {
      ${$maxItemCountPerPage && `max-height: ${$maxItemCountPerPage * itemToken.height[$size!]}px`};
      ${$hasScrollBar &&
        css`
          padding-right: ${wrapperToken.innerPaddingRight}px;
          ${scrollBarStyle(props)};
        `}
    }

    ${StyledBaseMenuItemDivider} {
      background: ${colors.pattern.background.divider};
      margin: ${dividerToken.paddingVertical}px ${$hasScrollBar ? 0 : dividerToken.paddingRight}px
        ${dividerToken.paddingVertical}px ${dividerToken.paddingLeft[$size!]}px;
    }

    ${StyledBaseMenuItemGroup} {
      ${StyledBaseMenuItem} {
        padding-left: ${groupToken.paddingLeft[$size!] + groupToken.itemIndent}px;
      }
    }

    ${StyledBaseMenuItemGroupLabel} {
      font-size: ${groupToken.fontSize}px;
      line-height: ${groupToken.height}px;
      height: ${groupToken.height}px;
      background: ${groupToken.background};
      color: ${groupToken.color};
      padding-left: ${groupToken.paddingLeft[$size!]}px;
    }
  `;
};

export const StyledBaseMenu = styled.div`
  ${baseMenuCss};
`;
