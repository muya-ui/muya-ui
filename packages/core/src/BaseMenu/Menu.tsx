import React, { Ref, useMemo } from 'react';

import useTheme from '../utils/useTheme';
import { StyledBaseMenu, StyledBaseMenuWrapper } from './styled';
import { IBaseMenuProps } from './types';
import { getMenuItemCountFromChildren } from './utils';

const Menu = React.forwardRef((props: IBaseMenuProps, ref: Ref<HTMLDivElement>) => {
  const { maxItemCountPerPage, width, size, children, onScroll, ...restProps } = props;
  const theme = useTheme();
  const hasScrollBar = useMemo(() => {
    if (maxItemCountPerPage) {
      let count = getMenuItemCountFromChildren(children);
      return count > maxItemCountPerPage;
    } else {
      return false;
    }
  }, [children, maxItemCountPerPage]);
  const wrapperWidth = useMemo(() => {
    if (width && typeof width === 'object') {
      return width[size!];
    } else {
      return width;
    }
  }, [size, width]);

  return (
    <StyledBaseMenu
      ref={ref}
      theme={theme}
      $size={size}
      $width={wrapperWidth}
      $hasScrollBar={hasScrollBar}
      $maxItemCountPerPage={maxItemCountPerPage}
      {...restProps}
    >
      <StyledBaseMenuWrapper onScroll={onScroll}>{children}</StyledBaseMenuWrapper>
    </StyledBaseMenu>
  );
});

Menu.defaultProps = {
  size: 'm',
};

export default React.memo(Menu);
