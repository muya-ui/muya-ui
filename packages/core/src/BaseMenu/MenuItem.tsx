import React, { cloneElement, Ref, useMemo, useEffect, useRef } from 'react';
import { isElement } from 'react-is';

import { SelectIcon } from '@muya-ui/theme-light';

import Spin from '../Spin';
import memoForwardRef from '../utils/memoForwardRef';
import useTheme from '../utils/useTheme';
import { StyledBaseMenuItem } from './styled';
import { IBaseMenuItemProps } from './types';
import scrollIntoView from '../utils/scrollIntoView';
import { useForkRef } from '@muya-ui/utils';

const MenuItem = memoForwardRef((props: IBaseMenuItemProps, ref: Ref<HTMLDivElement>) => {
  const {
    size = 'm',
    children,
    selected = false,
    disabled = false,
    active = false,
    loading = false,
    loadingIcon,
    selectedIcon,
    ...restProps
  } = props;
  const theme = useTheme();
  const itemRef = useRef<HTMLDivElement>(null);
  const handleRef = useForkRef(ref, itemRef);
  const {
    components: { BaseMenu: token },
  } = theme;

  let suffixIconNode;
  let SuffixIcon;
  if (loading) {
    suffixIconNode = loadingIcon || (
      <Spin
        style={{
          flexShrink: 0,
          marginLeft: token.iconMarginLeft,
          color: token.iconColor,
        }}
      />
    );
  } else if (selected) {
    SuffixIcon = token.selectedIcon || SelectIcon;
    suffixIconNode = selectedIcon || (
      <SuffixIcon
        style={{
          flexShrink: 0,
          marginLeft: token.iconMarginLeft,
          color: token.iconColor,
        }}
      />
    );
  }

  const finalChildren = useMemo(
    () =>
      isElement(children) ? (
        cloneElement(children, {
          size,
          theme,
        })
      ) : (
        <span>{children}</span>
      ),
    [children, size, theme],
  );

  useEffect(() => {
    if (active && itemRef.current) {
      scrollIntoView(itemRef.current, {
        onlyFirstScrollableParent: true,
        behavior: 'auto',
        isScrollable: (element: HTMLElement) => getComputedStyle(element).overflowY === 'auto',
        align: {
          top: 0.5,
        },
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [active]);

  return (
    <StyledBaseMenuItem
      ref={handleRef}
      theme={theme}
      $size={size}
      $active={active}
      $selected={selected}
      $disabled={disabled}
      {...restProps}
    >
      {finalChildren}
      {suffixIconNode}
    </StyledBaseMenuItem>
  );
});

MenuItem.displayName = 'MenuItem';

export default MenuItem;
