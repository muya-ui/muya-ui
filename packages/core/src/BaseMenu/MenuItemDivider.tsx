import { StyledBaseMenuItemDivider } from './styled';

const MenuItemDivider = StyledBaseMenuItemDivider;

MenuItemDivider.displayName = 'MenuItemDivider';

export default MenuItemDivider;
