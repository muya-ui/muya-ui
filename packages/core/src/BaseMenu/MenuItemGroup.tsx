import React from 'react';

import { StyledBaseMenuItemGroup, StyledBaseMenuItemGroupLabel } from './styled';
import { IBaseMenuItemGroupProps } from './types';

const MenuItemGroup = React.memo((props: IBaseMenuItemGroupProps) => {
  const { label, children, ...restProps } = props;
  return (
    <StyledBaseMenuItemGroup {...restProps}>
      <StyledBaseMenuItemGroupLabel>{label}</StyledBaseMenuItemGroupLabel>
      {children}
    </StyledBaseMenuItemGroup>
  );
});

MenuItemGroup.displayName = 'MenuItemGroup';

export default MenuItemGroup;
