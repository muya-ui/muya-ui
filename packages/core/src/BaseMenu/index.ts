import Menu from './Menu';
import MenuItem from './MenuItem';
import MenuItemDivider from './MenuItemDivider';
import MenuItemGroup from './MenuItemGroup';

const BaseMenu = {
  Menu: Menu,
  Item: MenuItem,
  Group: MenuItemGroup,
  Divider: MenuItemDivider,
};

export default BaseMenu;

export * from './types';
