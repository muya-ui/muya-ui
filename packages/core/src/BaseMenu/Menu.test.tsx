import React from 'react';
import renderer from 'react-test-renderer';

import BaseMenu from './index';

test('测试普通渲染', () => {
  const tree = renderer
    .create(
      <BaseMenu.Menu maxItemCountPerPage={5}>
        <BaseMenu.Group label="分组一">
          <BaseMenu.Item>选项一</BaseMenu.Item>
          <BaseMenu.Item>选项二</BaseMenu.Item>
        </BaseMenu.Group>
        <BaseMenu.Divider></BaseMenu.Divider>
        <BaseMenu.Item loading>选项普通</BaseMenu.Item>
        <BaseMenu.Item selected>选项选中</BaseMenu.Item>
        <BaseMenu.Item disabled>选项禁用</BaseMenu.Item>
        <BaseMenu.Item loading>选项加载</BaseMenu.Item>
      </BaseMenu.Menu>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试不同宽度渲染', () => {
  const tree = renderer
    .create(
      <BaseMenu.Menu
        width={{
          s: 100,
          m: 200,
          l: 300,
          xl: 400,
        }}
      >
        <BaseMenu.Group label="分组一">
          <BaseMenu.Item>选项一</BaseMenu.Item>
          <BaseMenu.Item>选项二</BaseMenu.Item>
        </BaseMenu.Group>
        <BaseMenu.Divider></BaseMenu.Divider>
        <BaseMenu.Item loading>选项普通</BaseMenu.Item>
        <BaseMenu.Item selected>选项选中</BaseMenu.Item>
        <BaseMenu.Item disabled>选项禁用</BaseMenu.Item>
        <BaseMenu.Item loading>选项加载</BaseMenu.Item>
      </BaseMenu.Menu>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
