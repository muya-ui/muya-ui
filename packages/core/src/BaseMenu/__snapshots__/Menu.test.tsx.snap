// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试不同宽度渲染 1`] = `
.c11 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
  padding: 0 12px;
  font-size: 12px;
  line-height: 32px;
  height: 32px;
  cursor: pointer;
  -webkit-transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  color: rgba(102,102,102,1);
  background-color: rgba(255,255,255,1);
  font-weight: 400;
}

.c7:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c7:active {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c7 span {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
  padding: 0 12px;
  font-size: 12px;
  line-height: 32px;
  height: 32px;
  cursor: pointer;
  -webkit-transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  color: rgba(102,102,102,1);
  background-color: rgba(255,255,255,1);
  font-weight: 400;
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.04);
  font-weight: 600;
}

.c13:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c13:active {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c13 span {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c13:hover {
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.08);
}

.c13:active {
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.04);
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
  padding: 0 12px;
  font-size: 12px;
  line-height: 32px;
  height: 32px;
  cursor: pointer;
  -webkit-transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  color: rgba(102,102,102,1);
  background-color: rgba(255,255,255,1);
  font-weight: 400;
  color: rgba(221,221,221,1);
  background-color: rgba(255,255,255,1);
  font-weight: 400;
  cursor: not-allowed;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c14:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c14:active {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c14 span {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c14:hover {
  color: rgba(221,221,221,1);
  background-color: rgba(255,255,255,1);
}

.c14:active {
  color: rgba(221,221,221,1);
  background-color: rgba(255,255,255,1);
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  cursor: default;
}

.c9 {
  height: 1px;
}

.c2 {
  overflow-y: auto;
}

.c15 .c1 {
  max-height: 160px;
  padding-right: 4px;
}

.c15 .c1::-webkit-scrollbar {
  width: 4px;
  height: 4px;
}

.c15 .c1::-webkit-scrollbar-thumb {
  border-radius: 999rem;
  background-color: transparent;
}

.c15 .c1:hover::-webkit-scrollbar-thumb {
  background-color: rgba(51,51,51,0.12);
}

.c15 .c1:hover::-webkit-scrollbar-thumb:hover {
  background-color: rgba(51,51,51,0.16);
}

.c15 .c8 {
  background: rgba(238,238,238,1);
  margin: 5.5px 0px 5.5px 12px;
}

.c15 .c3 .c6 {
  padding-left: 24px;
}

.c15 .c4 {
  font-size: 12px;
  line-height: 24px;
  height: 24px;
  background: rgba(255,255,255,1);
  color: rgba(187,187,187,1);
  padding-left: 12px;
}

.c0 {
  box-shadow: 0 0 12px 0 rgba(102,102,102,0.12);
  background: rgba(255,255,255,1);
  border-radius: 2px;
  padding: 4px 0;
  overflow: hidden;
  width: 200px;
}

.c0 .c8 {
  background: rgba(238,238,238,1);
  margin: 5.5px 12px 5.5px 12px;
}

.c0 .c3 .c6 {
  padding-left: 24px;
}

.c0 .c4 {
  font-size: 12px;
  line-height: 24px;
  height: 24px;
  background: rgba(255,255,255,1);
  color: rgba(187,187,187,1);
  padding-left: 12px;
}

.c12 {
  -webkit-animation: iECmZH 0.75s linear infinite,keNNsN 1.5s linear infinite;
  animation: iECmZH 0.75s linear infinite,keNNsN 1.5s linear infinite;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="c0"
>
  <div
    className="c1 c2"
  >
    <div
      className="c3 "
    >
      <span
        className="c4 c5"
      >
        分组一
      </span>
      <div
        className="c6 c7"
      >
        <span>
          选项一
        </span>
      </div>
      <div
        className="c6 c7"
      >
        <span>
          选项二
        </span>
      </div>
    </div>
    <div
      className="c8 c9"
    />
    <div
      className="c6 c7"
    >
      <span>
        选项普通
      </span>
      <span
        className="c10"
        direction="column"
      >
        <span
          className=""
        >
          <svg
            aria-hidden="false"
            className="c11 c12"
            focusable="false"
            preserveAspectRatio="xMidYMid"
            style={
              Object {
                "color": "inherit",
                "flexShrink": 0,
                "marginLeft": 12,
              }
            }
            viewBox="0 0 100 100"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="45"
              stroke="rgba(26,122,248,1)"
              strokeLinecap="round"
              strokeWidth="10"
              transform="rotate(144 50 50)"
            />
          </svg>
        </span>
        <span />
      </span>
    </div>
    <div
      className="c6 c13"
    >
      <span>
        选项选中
      </span>
      <svg
        aria-hidden="false"
        className="c11"
        focusable="false"
        style={
          Object {
            "color": "inherit",
            "flexShrink": 0,
            "marginLeft": 12,
          }
        }
        viewBox="0 0 1024 1024"
      >
        <path
          d="M407.787 619.136l454.517-448.47 119.03 116.96-558.4 550.945a21.333 21.333 0 01-29.931.032L42.667 494.1l118.826-117.16 246.294 242.197z"
        />
      </svg>
    </div>
    <div
      className="c6 c14"
    >
      <span>
        选项禁用
      </span>
    </div>
    <div
      className="c6 c7"
    >
      <span>
        选项加载
      </span>
      <span
        className="c10"
        direction="column"
      >
        <span
          className=""
        >
          <svg
            aria-hidden="false"
            className="c11 c12"
            focusable="false"
            preserveAspectRatio="xMidYMid"
            style={
              Object {
                "color": "inherit",
                "flexShrink": 0,
                "marginLeft": 12,
              }
            }
            viewBox="0 0 100 100"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="45"
              stroke="rgba(26,122,248,1)"
              strokeLinecap="round"
              strokeWidth="10"
              transform="rotate(144 50 50)"
            />
          </svg>
        </span>
        <span />
      </span>
    </div>
  </div>
</div>
`;

exports[`测试普通渲染 1`] = `
.c11 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
  padding: 0 12px;
  font-size: 12px;
  line-height: 32px;
  height: 32px;
  cursor: pointer;
  -webkit-transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  color: rgba(102,102,102,1);
  background-color: rgba(255,255,255,1);
  font-weight: 400;
}

.c7:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c7:active {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c7 span {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
  padding: 0 12px;
  font-size: 12px;
  line-height: 32px;
  height: 32px;
  cursor: pointer;
  -webkit-transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  color: rgba(102,102,102,1);
  background-color: rgba(255,255,255,1);
  font-weight: 400;
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.04);
  font-weight: 600;
}

.c13:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c13:active {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c13 span {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c13:hover {
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.08);
}

.c13:active {
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.04);
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
  padding: 0 12px;
  font-size: 12px;
  line-height: 32px;
  height: 32px;
  cursor: pointer;
  -webkit-transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  transition: background-color 150ms cubic-bezier(0.4,0,0.2,1),color 150ms cubic-bezier(0.4,0,0.2,1);
  color: rgba(102,102,102,1);
  background-color: rgba(255,255,255,1);
  font-weight: 400;
  color: rgba(221,221,221,1);
  background-color: rgba(255,255,255,1);
  font-weight: 400;
  cursor: not-allowed;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c14:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c14:active {
  color: rgba(102,102,102,1);
  background-color: rgba(250,250,250,1);
}

.c14 span {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c14:hover {
  color: rgba(221,221,221,1);
  background-color: rgba(255,255,255,1);
}

.c14:active {
  color: rgba(221,221,221,1);
  background-color: rgba(255,255,255,1);
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  cursor: default;
}

.c9 {
  height: 1px;
}

.c2 {
  overflow-y: auto;
}

.c0 {
  box-shadow: 0 0 12px 0 rgba(102,102,102,0.12);
  background: rgba(255,255,255,1);
  border-radius: 2px;
  padding: 4px 0;
  overflow: hidden;
  padding-right: 4px;
}

.c0 .c1 {
  max-height: 160px;
  padding-right: 4px;
}

.c0 .c1::-webkit-scrollbar {
  width: 4px;
  height: 4px;
}

.c0 .c1::-webkit-scrollbar-thumb {
  border-radius: 999rem;
  background-color: transparent;
}

.c0 .c1:hover::-webkit-scrollbar-thumb {
  background-color: rgba(51,51,51,0.12);
}

.c0 .c1:hover::-webkit-scrollbar-thumb:hover {
  background-color: rgba(51,51,51,0.16);
}

.c0 .c8 {
  background: rgba(238,238,238,1);
  margin: 5.5px 0px 5.5px 12px;
}

.c0 .c3 .c6 {
  padding-left: 24px;
}

.c0 .c4 {
  font-size: 12px;
  line-height: 24px;
  height: 24px;
  background: rgba(255,255,255,1);
  color: rgba(187,187,187,1);
  padding-left: 12px;
}

.c12 {
  -webkit-animation: iECmZH 0.75s linear infinite,keNNsN 1.5s linear infinite;
  animation: iECmZH 0.75s linear infinite,keNNsN 1.5s linear infinite;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="c0"
>
  <div
    className="c1 c2"
  >
    <div
      className="c3 "
    >
      <span
        className="c4 c5"
      >
        分组一
      </span>
      <div
        className="c6 c7"
      >
        <span>
          选项一
        </span>
      </div>
      <div
        className="c6 c7"
      >
        <span>
          选项二
        </span>
      </div>
    </div>
    <div
      className="c8 c9"
    />
    <div
      className="c6 c7"
    >
      <span>
        选项普通
      </span>
      <span
        className="c10"
        direction="column"
      >
        <span
          className=""
        >
          <svg
            aria-hidden="false"
            className="c11 c12"
            focusable="false"
            preserveAspectRatio="xMidYMid"
            style={
              Object {
                "color": "inherit",
                "flexShrink": 0,
                "marginLeft": 12,
              }
            }
            viewBox="0 0 100 100"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="45"
              stroke="rgba(26,122,248,1)"
              strokeLinecap="round"
              strokeWidth="10"
              transform="rotate(144 50 50)"
            />
          </svg>
        </span>
        <span />
      </span>
    </div>
    <div
      className="c6 c13"
    >
      <span>
        选项选中
      </span>
      <svg
        aria-hidden="false"
        className="c11"
        focusable="false"
        style={
          Object {
            "color": "inherit",
            "flexShrink": 0,
            "marginLeft": 12,
          }
        }
        viewBox="0 0 1024 1024"
      >
        <path
          d="M407.787 619.136l454.517-448.47 119.03 116.96-558.4 550.945a21.333 21.333 0 01-29.931.032L42.667 494.1l118.826-117.16 246.294 242.197z"
        />
      </svg>
    </div>
    <div
      className="c6 c14"
    >
      <span>
        选项禁用
      </span>
    </div>
    <div
      className="c6 c7"
    >
      <span>
        选项加载
      </span>
      <span
        className="c10"
        direction="column"
      >
        <span
          className=""
        >
          <svg
            aria-hidden="false"
            className="c11 c12"
            focusable="false"
            preserveAspectRatio="xMidYMid"
            style={
              Object {
                "color": "inherit",
                "flexShrink": 0,
                "marginLeft": 12,
              }
            }
            viewBox="0 0 100 100"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="45"
              stroke="rgba(26,122,248,1)"
              strokeLinecap="round"
              strokeWidth="10"
              transform="rotate(144 50 50)"
            />
          </svg>
        </span>
        <span />
      </span>
    </div>
  </div>
</div>
`;
